src/compiler/core.ts

```diff
@@ -162,6 +165,16 @@ namespace ts {
         return true;
     }
 
+    /**
+     * Equivalent to every, but can discriminate between unions of arrays.
+     */
+    export function isArrayOf<T>(array: readonly any[], callback: (element: any, index: number) => element is T): array is readonly T[];

```

Comment:
I don't like that I had to add this helper, but I cannot make `every` do the right narrowing because inference always seems to pick the wrong type for `T`.

Reply:
Removed in favor of just telling `every` what type it should be dealing with, which is precented by other things like uses of `find` or similar.